cover.c

```diff
@@ -150,9 +154,97 @@ void cover_dump(CoverList* cover, FILE* fp) {
         fprintf(fp, "%d", j);
       }
     }
-    fprintf(fp, "]}");
+
+    fprintf(fp, "],\"phases\":{");
+
+        fprintf(fp, "\"BEGIN\":[");
+        lcount = 0;
+        for (j = 1; j <= node->bmax; ++j) {
+          if (LINE_IS_PRESENT_OR_COVERED(node->lines, j)) {
+            if (LINE_GET_COMPILER_PHASE(node->phases, j) <= PERL_PHASE_START ) {
+              if (lcount++) {
+                fprintf(fp, ",");
+              }
+              fprintf(fp, "%d", j);
+            }
+          }
+        }
+        fprintf(fp, "],");
+
+        fprintf(fp, "\"CHECK\":[");
+        lcount = 0;
+        for (j = 1; j <= node->bmax; ++j) {
+          if (LINE_IS_PRESENT_OR_COVERED(node->lines, j)) {
+            if (LINE_GET_COMPILER_PHASE(node->phases, j) == PERL_PHASE_CHECK ) {
+              if (lcount++) {
+                fprintf(fp, ",");
+              }
+              fprintf(fp, "%d", j);
+            }
+          }
+        }
+        fprintf(fp, "],");
+
+        fprintf(fp, "\"INIT\":[");
+        lcount = 0;
+        for (j = 1; j <= node->bmax; ++j) {
+          if (LINE_IS_PRESENT_OR_COVERED(node->lines, j)) {
+            if (LINE_GET_COMPILER_PHASE(node->phases, j) == PERL_PHASE_INIT ) {
+              if (lcount++) {
+                fprintf(fp, ",");
+              }
+              fprintf(fp, "%d", j);
+            }
+          }
+        }
+        fprintf(fp, "],");
+
+        fprintf(fp, "\"RUN\":[");
+        lcount = 0;
+        for (j = 1; j <= node->bmax; ++j) {
+          if (LINE_IS_PRESENT_OR_COVERED(node->lines, j)) {
+            if (LINE_GET_COMPILER_PHASE(node->phases, j) == PERL_PHASE_RUN ) {
+              if (lcount++) {
+                fprintf(fp, ",");
+              }
+              fprintf(fp, "%d", j);
+            }
+          }
+        }
+        fprintf(fp, "],");
+
+        fprintf(fp, "\"END\":[");
+        lcount = 0;
+        for (j = 1; j <= node->bmax; ++j) {
+          if (LINE_IS_PRESENT_OR_COVERED(node->lines, j)) {
+            if (LINE_GET_COMPILER_PHASE(node->phases, j) == PERL_PHASE_END ) {
+              if (lcount++) {
+                fprintf(fp, ",");
+              }
+              fprintf(fp, "%d", j);
+            }
+          }
+        }
+        fprintf(fp, "],");
+
+        fprintf(fp, "\"DESTRUCT\":[");
+        lcount = 0;
+        for (j = 1; j <= node->bmax; ++j) {
+          if (LINE_IS_PRESENT_OR_COVERED(node->lines, j)) {
+            if (LINE_GET_COMPILER_PHASE(node->phases, j) == PERL_PHASE_DESTRUCT ) {
+              if (lcount++) {
+                fprintf(fp, ",");
+              }
+              fprintf(fp, "%d", j);
+            }
+          }
+        }
+        fprintf(fp, "]");
+
+    fprintf(fp, "}"); /* close the `phases` object */
```

Comment:
The above truly is horrible, probably best if I create a macro or something. Comments?